### PR TITLE
Bug 1953077: rework GCP passthrough permissions checking

### PR DESCRIPTION
**Pull request by joelddiaz**
In GCP Passthrough mode, the permissions are dynamically checked before
copying the secret contents over. This is done by performing a
TestIamPermissions() call against the current GCP project.

There are certain permissions that make no sense to check against the
project level. Periodically fetch a list of testable permissions using the
QueryTestablePermissions() API call.

Before making a TestIamPermissions() call, filter the permissions through the
allowed list to avoid needlessly receiving errors when calling TestIamPermissions()
against a GCP Project resource.

### CODE REVIEW

**Comment by joelddiaz** [comment]
/hold
we should discuss what this means for CredentialsRequests we don't control. A user can cause CCO to perform many, many API calls as it cycles through TestIamPermissions() to find 1 rejection, remove the rejected permission, and run TestIamPermissions() again, and again for each rejected permission.

File: pkg/operator/utils/gcp/utils.go [hunk 1]
```diff
@@ -164,22 +123,57 @@ func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []strin
...
-func filterOutPermissions(permList, filterOutList []string) []string {
-	filteredPerms := []string{}
-	for _, perm := range permList {
-		filterOut := false
-		for _, filterPerm := range filterOutList {
-			if perm == filterPerm {
-				filterOut = true
+func refreshTestablePermissions(gcpClient ccgcp.Client, projectName string, logger log.FieldLogger) error {
+	// refresh if last update was over 1 hour ago
+	if testablePerms.lastUpdated.Add(time.Hour).Before(time.Now()) {
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:128]: early return maybe?

File: pkg/operator/utils/gcp/utils.go [hunk 2]
```diff
@@ -164,22 +123,57 @@ func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []strin
...
+
+		ctx := context.TODO()
+		nextPageToken := ""
+		permList := []string{}
+
+		for {
+			projectNamePath := fmt.Sprintf(`//cloudresourcemanager.googleapis.com/projects/%s`, projectName)
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:135]: this can be outside the loop

File: pkg/operator/utils/gcp/utils.go [hunk 3]
```diff
@@ -164,22 +123,57 @@ func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []strin
...
+			req := &iamadminpb.QueryTestablePermissionsRequest{
+				FullResourceName: projectNamePath,
+				PageToken:        nextPageToken,
+			}
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:139]: same let's create the req body once before loop, and only update the PageToken

File: pkg/operator/utils/gcp/utils.go [hunk 4]
```diff
@@ -164,22 +123,57 @@ func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []strin
...
+			resp, err := gcpClient.QueryTestablePermissions(ctx, req)
+			if err != nil {
+				logger.WithError(err).Error("failed to gather list of testable permissions")
+				return err
+			}
+
+			for _, perm := range resp.Permissions {
+				permList = append(permList, perm.Name)
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:147]: just directly add to the new local copy of set we can generate.

File: pkg/operator/utils/gcp/utils.go [hunk 5]
```diff
@@ -164,22 +123,57 @@ func checkServicesAndPermissions(gcpClient ccgcp.Client, permissionsList []strin
...
+	}
+
+	return nil
+}
+
+// filterOutPermissions will take a list of GCP permissions to test against and return a list of permissions that
+// are valid to test against at the project level
+func filterOutPermissions(gcpClient ccgcp.Client, projectName string, permList []string, logger log.FieldLogger) []string {
+
+	refreshTestablePermissions(gcpClient, projectName, logger)
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:168]: handle error.

Comment (joelddiaz) [reply pkg/operator/utils/gcp/utils.go]: last second refactoring (－‸ლ)

**Comment by joelddiaz** [comment]
@abhinavdahiya this should be good for another review

File: pkg/operator/utils/gcp/utils.go [hunk 6]
```diff
@@ -183,6 +227,51 @@ func CheckPermissionsAgainstPermissionList(gcpClient ccgcp.Client, permList []st
 	return disallowedPerms.Len() == 0, nil
 }
 
+/*
```
Comment (abhinavdahiya) [inline pkg/operator/utils/gcp/utils.go:230]: can we drop this commented code?

Comment (joelddiaz) [reply pkg/operator/utils/gcp/utils.go]: Good catch. Done.

**Pull request merged**